### ISSUE TITLE
Implementing the wayland backend

### ISSUE THREAD

**Issue by vegardsjo** [treeform/windy#113]
Hi,

I've been hacking away on the wayland backend part of windy. I'm working on adding support for hardware graphics buffers as well as supporting some of the wayland protocol extensions used in modern wayland desktops, like:

- https://wayland.app/protocols/xdg-shell
  Standard shell surface (window) management

- https://wayland.app/protocols/idle-inhibit-unstable-v1
  Apps can tell the compositor not go to sleep/dpms if they are "active". This is basically "prevent screensaver" for wayland.
  It's needed for apps/games using input methods not handled by the compositor's idle system.
  This can even be regular gamepads if the compositor doesn't let the inputs of such a device "reset the idle timer"

- https://wayland.app/protocols/linux-dmabuf-unstable-v1
  Use graphics buffers coming from anywhere, e.g. from a camera, but in general it brings much flexibility to the table. 
  Also needed or very useful for certain embedded systems (and accursed (nvidia) drivers).
  Also useful in order to do single-buffering on wayland, if one wants that.

- https://wayland.app/protocols/presentation-time
  Needed for video/audio synchronization

- https://wayland.app/protocols/wlr-layer-shell-unstable-v1
  So windy can be used to build bars/menus/launchers for wayland based desktops :grinning:

Are you open for contributions/PRs/discussions? :)

**Comment by treeform** [comment]
Yes we are. We already have some wayland started, but it will be nice to have that finished and passing the tests.

Linux desktop is hard. X11 and Wayland are great but they don't give you Dialogs, Tray icons or IME. One needs to also integrate with GTK or QT to get that.

**Comment by xTrayambak** [comment]
I second this! It'd be so convenient if windy could be a native-Wayland library as a ton of visual glitches can happen when passing windy-based apps through XWayland like flickering, random black screens and sometimes, the compositor itself crashes!

**Comment by vegardsjo** [comment]
@treeform Hi again, I've made good progress and think this will be ready for a PR in the coming weeks, but I have a bunch of api design and usage related questions as well as some related to boxy and pixie. Could we discuss this somewhere more instant (e.g. IRC or IM)?

**Comment by xTrayambak** [comment]
Hey, @vegardsjo, can you please make your fork a repo so I can contribute to it as well? I've written my own "solution" for a Wayland backend but it's fairly awful. Let me know if you need any help.

**Comment by treeform** [comment]
@vegardsjo I am available on Nim's discord channels: https://discord.com/invite/ezDFDw2 send me a DM.

**Comment by vegardsjo** [comment]
> Hey, @vegardsjo, can you please make your fork a repo so I can contribute to it as well? I've written my own "solution" for a Wayland backend but it's fairly awful. Let me know if you need any help.

The code's all over the place right now, but I'll let you know when I get the time to clean it up and push my fork/branch :+1:

**Comment by xTrayambak** [comment]
@treeform I'm currently working on a high-level Wayland wrapper for Nim. Would you be down to accept a MR that implements this with that?
